src/Common/MemoryWorker.cpp

```diff
@@ -310,16 +311,34 @@ void MemoryWorker::backgroundThread()
             ProfileEvents::increment(ProfileEvents::MemoryAllocatorPurge);
             ProfileEvents::increment(ProfileEvents::MemoryAllocatorPurgeTimeMicroseconds, purge_watch.elapsedMicroseconds());
         }
-#endif
 
-#if USE_JEMALLOC
+        /// update MemoryTracker with `allocated` information from jemalloc when:
+        ///  - it's a first run of MemoryWorker (MemoryTracker could've missed some allocation before its initialization)
+        ///  - MemoryTracker stores a negative value
+        ///  - `correct_tracker` is set to true
         if (unlikely(first_run || total_memory_tracker.get() < 0))
         {
             if (source != MemoryUsageSource::Jemalloc)
                 epoch_mib.setValue(0);
 
-            MemoryTracker::updateAllocated(allocated_mib.getValue());
+            MemoryTracker::updateAllocated(allocated_mib.getValue(), /*log_change=*/true);
         }
+        else if (correct_tracker)
+        {
+            if (source != MemoryUsageSource::Jemalloc)
+                epoch_mib.setValue(0);
+
+            MemoryTracker::updateAllocated(allocated_mib.getValue(), /*log_change=*/false);
+        }
+#else
+        /// we don't update in the first run if we don't have jemalloc
+        /// because we can only use resident memory information
+        /// resident memory can be much larger than the actual allocated memory
+        /// so we rather ignore the potential difference caused by allocated memory
+        /// before MemoryTracker initialization
+        if (unlikely(total_memory_tracker.get() < 0) || correct_tracker)
```

Comment:
I doubt that it makes any difference (updating for the first run or not), but anyway this is minor